### PR TITLE
working new item button for nested (> 2nd level)

### PR DESCRIPTION
**Pull request by sakarikl**
Basically same functionality as in 
https://github.com/sonata-project/SonataAdminBundle/pull/2985 but BC breaks fixed

fixed also to work with underscore model properties
https://github.com/sonata-project/SonataAdminBundle/pull/2985#issuecomment-150670027

### CODE REVIEW

**Comment by osavchenko** [comment]
As for me looks great but could you fix your PR according to notices from build?

**Comment by sakarikl** [comment]
at least part of the build errors are following:

/home/travis/build/sonata-project/SonataAdminBundle/Resources/doc/cookbook/recipe_sortable_listing.rst:50: WARNING: Could not parse literal_block as "yaml". highlighting skipped.

Which have nothing to do with this PR.

**Comment by OskarStark** [comment]
@sakarikl can you please add some tests?

**Comment by greg0ire** [comment]
see #3557

**Comment by greg0ire** [comment]
Please rebase on the latest commit, it should fix the build.

**Comment by greg0ire** [comment]
Please fixup the "oops, I forgot something, I need to fix cs" commits with the first commit. We don't need to know that you made some typo. Also, there is a merge commit in your P.R. , it is not normal.

File: Admin/AdminHelper.php [hunk 1]
```diff
@@ -201,4 +229,103 @@ public function camelize($property)
...
+            }
+        }
+
+        return $return_value;
+    }
+
+    /**
+     * check if given path exists in $entity.
+     *
+     * @param $propertyAccessor
```
Comment (OskarStark) [inline Admin/AdminHelper.php]: please complete the phpdoc with the classname

File: Admin/AdminHelper.php [hunk 2]
```diff
@@ -201,4 +229,103 @@ public function camelize($property)
...
+     * @param mixed  $entity
+     * @param string $path
+     *
+     * @return bool
```
Comment (OskarStark) [inline Admin/AdminHelper.php]: could also throw an exception, please add to the docs

File: Admin/AdminHelper.php [hunk 3]
```diff
@@ -201,4 +229,103 @@ public function camelize($property)
     {
         return BaseFieldDescription::camelize($property);
     }
+
+    /**
+     * @param AdminInterface $admin
+     * @param $elements
```
Comment (OskarStark) [inline Admin/AdminHelper.php]: which type?

**Comment by OskarStark** [comment]
@sakarikl can you please apply my comments?

@greg0ire can you please merge this PR afterwards? Thank you :+1:

**Comment by greg0ire** [comment]
@OskarStark : I do not have merge rights on this repository, sorry.

File: Admin/AdminHelper.php [hunk 4]
```diff
@@ -201,4 +229,105 @@ public function camelize($property)
...
         return BaseFieldDescription::camelize($property);
     }
+
+    /**
+     * @param AdminInterface $admin
+     * @param array          $elements
+     *
+     * @return string
+     */
+    protected function entityClassNameFinder(AdminInterface $admin, $elements)
```
Comment (greg0ire) [inline Admin/AdminHelper.php]: method names should be verbs, not names

**Comment by OskarStark** [comment]
Oh ok, i thought this because of the `Sonata Project member` label here

File: Admin/AdminHelper.php (hunk 4 above)
Comment (OskarStark) [inline Admin/AdminHelper.php]: :+1:

File: Admin/AdminHelper.php [hunk 5]
```diff
@@ -201,4 +229,105 @@ public function camelize($property)
     {
         return BaseFieldDescription::camelize($property);
     }
+
+    /**
```
Comment (greg0ire) [inline Admin/AdminHelper.php]: Could you add a description, like, "recursively find the class name of the admin responsible for the element at the end of an association chain" ? It's hard to figure out what this method does (I might be wrong about the description, BTW)

File: Admin/AdminHelper.php [hunk 6]
```diff
@@ -201,4 +229,105 @@ public function camelize($property)
...
+     *
+     * @return string
+     *
+     * @throws \Exception
+     */
+    public function getElementAccessPath($elementId, $entity)
+    {
+        $propertyAccessor = new PropertyAccessor();
+
+        $initial = preg_replace('#(_(\d+)_)#', '[$2]', implode('_', explode('_', substr($elementId, strpos($elementId, '_') + 1))));
```
Comment (greg0ire) [inline Admin/AdminHelper.php]: Intial what? Please describe what is the deal with the underscores in the parameter description (line 253)

File: Admin/AdminHelper.php [hunk 7]
```diff
@@ -201,4 +229,105 @@ public function camelize($property)
...
+     * @param string           $path
+     *
+     * @return bool
+     *
+     * @throws \RuntimeException
+     */
+    private function pathExists(PropertyAccessor $propertyAccessor, $entity, $path)
+    {
+        if (method_exists($propertyAccessor, 'isReadable')) {
+            return $propertyAccessor->isReadable($entity, $path);
```
Comment (greg0ire) [inline Admin/AdminHelper.php]: In which case doesn't this exist? Please indicate it so that we can remove the piece of code below when dropping compatibility with old versions of symfony.

**Comment by greg0ire** [comment]
@OskarStark : yeah it's a bit confusing, it is because I have rights on the AdminSearchBundle

File: Admin/AdminHelper.php [hunk 8]
```diff
@@ -113,42 +116,67 @@ public function appendFormFieldElement(AdminInterface $admin, $subject, $element
...
+            if (!isset($data[$childFormBuilder->getName()])) {
+                $data[$childFormBuilder->getName()] = array();
+            }
+
+            $objectCount = count($value);
+            $postCount = count($data[$childFormBuilder->getName()]);
+
+            $fields = array_keys($fieldDescription->getAssociationAdmin()->getFormFieldDescriptions());
+
+            // for now, not sure how to do that
```
Comment (greg0ire) [inline Admin/AdminHelper.php]: What do you mean exacty?

Comment (sakarikl) [inline Admin/AdminHelper.php]: That was in original code. indentation changed in that.

Comment (greg0ire) [inline Admin/AdminHelper.php]: Oh my bad, I did not even notice that

Comment (OskarStark) [inline Admin/AdminHelper.php]: but can we improve this anyway?

File: Admin/AdminHelper.php [hunk 9]
```diff
@@ -113,42 +116,67 @@ public function appendFormFieldElement(AdminInterface $admin, $subject, $element
...
-        // retrieve the posted data
-        $data = $admin->getRequest()->get($formBuilder->getName());
+            $collection = $propertyAccessor->getValue($entity, $path);
 
-        if (!isset($data[$childFormBuilder->getName()])) {
-            $data[$childFormBuilder->getName()] = array();
-        }
+            if ($collection instanceof ArrayCollection) {
+                $entityClassName = $this->getEntityClassName($admin, explode('.', preg_replace('#\[\d*?\]#', '', $path)));
+            } elseif ($collection instanceof \Doctrine\ORM\PersistentCollection) {
```
Comment (wodka) [inline Admin/AdminHelper.php]: missing the fix for mongodb
https://github.com/sonata-project/SonataAdminBundle/pull/2985#discussion-diff-31056069

```php

        if ($collection instanceof \Doctrine\ORM\PersistentCollection || $collection instanceof \Doctrine\ODM\MongoDB\PersistentCollection) {
            //since doctrine 2.4
            $entityClassName = $collection->getTypeClass()->getName();
        } elseif ($collection instanceof \Doctrine\Common\Collections\Collection) {
            $entityClassName = $this->entityClassNameFinder($admin, explode('.', preg_replace('#\[\d*?\]#', '', $path)));
        } else {
            return;
        }
```

with this change it should work for all possible mappers of doctrine (including the easy check for ORM and ODM)

Comment (sakarikl) [inline Admin/AdminHelper.php]: added mongodb fix

File: Admin/AdminHelper.php (hunk 8 above)
Comment (sakarikl) [inline Admin/AdminHelper.php]: probably, but shouldn't it be in separate commit / PR.

Comment (OskarStark) [inline Admin/AdminHelper.php]: yes :+1:

**Comment by OskarStark** [comment]
i think after the description we this is RTM @core23 @greg0ire

File: Admin/AdminHelper.php [hunk 10]
```diff
@@ -201,4 +228,113 @@ public function camelize($property)
...
+                    $dot = '';
+                }
+            }
+
+            if ($dot !== '.') {
+                throw new \Exception(sprintf('Could not get element id from %s Failing part: %s', $elementId, $sub_value));
+            }
+
+            //check if array access was in this location originally
+            preg_match("#$value\[(\d+)#", $initial_path, $matches);
```
Comment (OskarStark) [inline Admin/AdminHelper.php]: please rename to `$initialPath`

File: Admin/AdminHelper.php [hunk 11]
```diff
@@ -201,4 +228,113 @@ public function camelize($property)
...
+
+            if (isset($matches[1])) {
+                $part_return_value .= '['.$matches[1].']';
```
Comment (OskarStark) [inline Admin/AdminHelper.php]: please rename to `$partReturnValue`

File: Admin/AdminHelper.php [hunk 12]
```diff
@@ -201,4 +228,113 @@ public function camelize($property)
...
+
+            $return_value .= $return_value ? '.'.$part_return_value : $part_return_value;
+            $part_return_value = '';
+
+            if (isset($parts[$key + 1])) {
+                $current_entity = $propertyAccessor->getValue($entity, $return_value);
+            }
+        }
+
+        return $return_value;
```
Comment (OskarStark) [inline Admin/AdminHelper.php]: please rename to `$returnValue`

File: Admin/AdminHelper.php [hunk 13]
```diff
@@ -201,4 +228,113 @@ public function camelize($property)
...
+     * @param mixed            $entity
+     * @param string           $path
+     *
+     * @return bool
+     *
+     * @throws \RuntimeException
+     */
+    private function pathExists(PropertyAccessor $propertyAccessor, $entity, $path)
+    {
+        //sf2 <= 2.3 did not have isReadable method for PropertyAccessor
```
Comment (OskarStark) [inline Admin/AdminHelper.php]: not sure if need this @core23 @greg0ire , because we don't support SF 2.3 in the current version anymore

File: Tests/Admin/AdminHelperTest.php [hunk 14]
```diff
@@ -112,4 +112,63 @@ public function testAddNewInstanceInflector()
...
+
+        $this->assertSame('path_to_object[0].and.more', $path);
+    }
+
+    /**
+     * tests only so far that actual value/object is retrieved.
+     *
+     * @expectedException     Exception
+     * @expectedExceptionCode 0
+     * @expectedExceptionMessage unknown collection class
```
Comment (OskarStark) [inline Tests/Admin/AdminHelperTest.php]: indention is wrong

File: Tests/Admin/AdminHelperTest.php [hunk 15]
```diff
@@ -112,4 +112,63 @@ public function testAddNewInstanceInflector()
...
+        $sub2_object = $this->getMock('stdClass', array('getMore'));
+        $sub3_object = $this->getMock('stdClass', array('getFinalData'));
+        $dataMapper = $this->getMock('Symfony\Component\Form\DataMapperInterface');
+        $formFactory = $this->getMock('Symfony\Component\Form\FormFactoryInterface');
+        $eventDispatcher = $this->getMock('Symfony\Component\EventDispatcher\EventDispatcherInterface');
+        $formBuilder = new FormBuilder('test', get_class($simpleObject), $eventDispatcher, $formFactory);
+        $childFormBuilder = new FormBuilder('subObject', get_class($sub_object), $eventDispatcher, $formFactory);
+
+        $object->expects($this->atLeastOnce())->method('getSubObject')->will($this->returnValue(array($sub_object)));
+        $sub_object->expects($this->atLeastOnce())->method('getAnd')->will($this->returnValue($sub2_object));
```
Comment (OskarStark) [inline Tests/Admin/AdminHelperTest.php]: please rename `$sub_object` to `$subObject`

**Comment by OskarStark** [comment]
@sakarikl please fix the variable names, we don't use `snake_case`, we use `studyCase`, thank you :+1:

File: Admin/AdminHelper.php (hunk 13 above)
Comment (greg0ire) [inline Admin/AdminHelper.php]: I did not know that. In that case yeah, it can be removed.

Comment (sakarikl) [inline Admin/AdminHelper.php]: Travis still has tests for 2.3

**Comment by OskarStark** [comment]
Thank you @sakarikl

**Comment by nielvrom** [comment]
I'm working on version 2.3, how can I implement this fix?

**Comment by chalasr** [comment]
@sakarikl Thank you for fixing this old and important issue.

Known issues

- #262 
- #1228 
- #1327

:+1:

**Comment by develth** [comment]
This results in my case into a 

    unknown collection class

The Admin classes itself are working (1st lvl)
